packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs

```diff
@@ -3,118 +3,163 @@ use error_stack::{IntoReport, Report, Result, ResultExt};
-                    GROUP BY link_type_version_id
-                )
-                SELECT base_uri, "version", links FROM aggregated
-                INNER JOIN ids ON ids.version_id = aggregated.link_type_version_id
-                "#,
-                &[&identifier],
+async fn single_by_source_entity_id(
+    client: &(impl GenericClient + Sync),
+    source_entity_id: EntityId,
+) -> Result<impl Iterator<Item = (VersionedUri, Outgoing)> + Send, QueryError> {
+    Ok(client
+        .query(
+            r#"
+            SELECT base_uri, "version", target_entity_id
+            FROM links
+            INNER JOIN ids ON ids.version_id = links.link_type_version_id
+            WHERE active AND NOT multi and source_entity_id = $1
+            "#,
+            &[&source_entity_id],
+        )
+        .await
+        .into_report()
+        .change_context(QueryError)
+        .attach_printable(source_entity_id)?
+        .into_iter()
+        .map(|row| {
+            (
+                VersionedUri::new(row.get(0), row.get::<_, i64>(1) as u32),
+                Outgoing::Single(row.get(2)),
             )
-            .await
-            .into_report()
-            .change_context(QueryError)
-            .attach_printable(identifier)?
-            .into_iter()
-            .map(|row| (VersionedUri::new(row.get(0), row.get::<_, i64>(1) as u32), Outgoing::Multiple(row.get(2))));
+        }))
+}
 
-        let single_links = self
-            .client
-            .as_client()
-            .query(
-                r#"
-                SELECT base_uri, "version", target_entity_id
+async fn multi_by_source_entity_id(
+    client: &(impl GenericClient + Sync),
+    source_entity_id: EntityId,
+) -> Result<impl Iterator<Item = (VersionedUri, Outgoing)> + Send, QueryError> {
+    Ok(client
+        .query(
+            r#"
+            WITH aggregated as (
+                SELECT link_type_version_id, ARRAY_AGG(target_entity_id ORDER BY multi_order ASC) as links
                 FROM links
-                INNER JOIN ids ON ids.version_id = links.link_type_version_id
-                WHERE active AND NOT multi and source_entity_id = $1
-                "#,
-                &[&identifier],
+                WHERE active AND multi and source_entity_id = $1
+                GROUP BY link_type_version_id
             )
-            .await
-            .into_report()
-            .change_context(QueryError)
-            .attach_printable(identifier)?
-            .into_iter()
-            .map(|row| {
-                (
-                    VersionedUri::new(row.get(0), row.get::<_, i64>(1) as u32),
-                    Outgoing::Single(row.get(2)),
-                )
-            });
-
-        Ok(Links::new(multi_links.chain(single_links).collect()))
-    }
+            SELECT base_uri, "version", links FROM aggregated
+            INNER JOIN ids ON ids.version_id = aggregated.link_type_version_id
+            "#,
+            &[&source_entity_id],
```

Comment:
Here `parameter_list` can be utilized as well

Reply:
Same as above